pkg/server/container_create_windows.go

```diff
@@ -517,18 +513,57 @@ func (c *criService) generateContainerSpec(id string, sandboxID string, sandboxP
 		}
 	}
 
+	// hyperv pci devices are exposed as windows devices on the container spec
+	// given to hcsshim.
+	addOCIWindowsDevices(&g, rhcso, config.Devices)
+
 	return g.Config, nil
 }
 
-func addOCIWindowsDevices(g *generator, devs []*runtime.Device) {
+// getWindowsDeviceInfo is a helper function that returns a spec specified device's
+// prefix and device identifier. A prefix is any string before "://" in the spec
+// device's `HostPath`.
+func getWindowsDeviceInfo(hostPath string) (string, string) {
+	substrings := strings.SplitN(hostPath, "://", 2)
+	if len(substrings) <= 1 {
+		return "", ""
+	}
+	return substrings[0], substrings[1]
+}
+
+// validWindowsDeviceForIsolation is a helper function that returns if a given windows
+// device prefix is valid for the isolation type of the container.
+//
+// Hyper-V isolated containers only support the prefixes "gpu" and "vpci".
+//
+// Process isolated containers only support the prefixes "vpciLocationPath"
+// and "vpciClassGUID"
+//
+// Returns true if the device type is valid for the container, false otherwise
+func validWindowsDeviceForIsolation(prefix string, rhcso *runhcsoptions.Options) bool {
+	isolation := rhcso.SandboxIsolation
+	if isolation == runhcsoptions.Options_PROCESS {
+		return prefix == "vpciLocationPath" || prefix == "vpciClassGUID"
+	} else {
+		return prefix == "vpci" || prefix == "gpu"
+	}
+	return false
+}
+
+// addOCIWindowsDevices parses the devices field on the spec and creates corresponding
+// `WindowsDevice` on the container config if valid.
+func addOCIWindowsDevices(g *generator, rhcso *runhcsoptions.Options, devs []*runtime.Device) {
 	for _, device := range devs {
-		if strings.HasPrefix(device.HostPath, "gpu://") {
-			gpuID := strings.TrimPrefix(device.HostPath, "gpu://")
-			gpuDevice := runtimespec.WindowsDevice{
-				ID:     gpuID,
-				IDType: "gpu",
+		prefix, deviceIdentifier := getWindowsDeviceInfo(device.HostPath)
+		if prefix == "" || deviceIdentifier == "" {
+			continue
```

Comment:
It seems like we should log a warning when we skip any device entries, either due to this conditional, or if `validWindowsDeviceForIsolation` fails.

Reply:
Actually seems better to just return an error if we fail to handle any devices. If the user cared enough to pass something in here, they probably want to know it failed rather than silently ignoring it.

Reply:
Okay, one other thing occurred to me. Should we make all the logic in cri effectively just pass through the values as-is, and then if they are invalid we fail in the shim? That might make it easier in the future when we want to upstream our cri changes, so that we don't need to update upstream when we want to support a new device `IDType`. What do you think?

Reply:
Hmm, I see. So take out all restrictions on the prefix in cri. I like that idea, since that makes it easier for us to iterate if we need to add more support for a variety of devices that may need special casing like "gpu" without needing to update cri every time.

Reply:
Cool, let's do that and see if we run into any issues.

Reply:
Worked fine with the hcsshim changes :)

Reply:
Can we still return an error here? I assume if we can't parse the device string into prefix/id, then something is wrong for sure.

Reply:
Done.